### PR TITLE
rustbuild: Enable WebAssembly backend by default

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit alters how we compile LLVM by default enabling the WebAssembly
backend. This then also adds the wasm32-unknown-unknown target to get compiled
on the `cross` builder and distributed through rustup. Tests are not yet enabled
for this target but that should hopefully be coming soon!

### CODE REVIEW

**Comment by alexcrichton** [comment]
This is the successor of https://github.com/rust-lang/rust/pull/45905 where now that everything's in tree it's time to add some tests! This PR will ensure that the wasm32-unknown-unknown target does not regress (keeps building) and also enables all of our distributed compilers to build for wasm32 architectures by default. This PR also adds a rustup target, so after this lands it will be possible to do:

```
rustup target add wasm32-unknown-unknown
```

and then you're ready to experiment! After this PR I'll work on adding a builder that starts running some tests so we can get some regression testing there.

**Comment by kennytm** [comment]
LGTM.

@bors r+

**Comment by est31** [comment]
Doesn't this need LLVM 5.0? Or was that requirement dropped?

**Comment by est31** [comment]
Hmm I guess it never was a hard requirement really, but there was just no testing on it...

**Comment by alexcrichton** [comment]
@bors: r- 

turns out rustdoc for wasm is super broke, will fix.

@est31 yes for a bug-free implementation we'll probably need LLVM 5.0. I don't think there's any need to block on that-which-is-unlikely-to-ever-happen though

**Comment by alexcrichton** [comment]
re-r? @kennytm 

The support added for documenting all platforms when documenting the standard library was broken with wasm b/c it didn't share enough in common to compile the unix/windows "tiny shims". I ended up using `cfg_if!` to clean up the declarations (cut down on `#[cfg]` traffic) and then had to fix the "everybody loops" implementation in rustc to get that working.

**Comment by kennytm** [comment]
r=me after tidy fix.

```
[00:03:41] tidy error: /checkout/src/librustc_driver/lib.rs:568: line longer than 100 chars
[00:03:42] some tidy checks failed
```

**Comment by kennytm** [comment]
Cannot build libstd for `x86_64-unknown-redox`.

<details>

```
[00:57:37] error[E0432]: unresolved import `sys::net::Socket`
[00:57:37]   --> /checkout/src/libstd/sys/unix/ext/net.rs:39:5
[00:57:37]    |
[00:57:37] 39 | use sys::net::Socket;
[00:57:37]    |     ^^^^^^^^^^^^^^^^ no `Socket` in `sys::redox::net`
[00:57:37] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::AsRawFd` for type `process::ChildStdin`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:154:1
[00:57:38]     |
[00:57:38] 154 | / impl AsRawFd for process::ChildStdin {
[00:57:38] 155 | |     fn as_raw_fd(&self) -> RawFd {
[00:57:38] 156 | |         self.as_inner().fd().raw()
[00:57:38] 157 | |     }
[00:57:38] 158 | | }
[00:57:38]     | |_^ conflicting implementation for `process::ChildStdin`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 144 | / impl AsRawFd for process::ChildStdin {
[00:57:38] 145 | |     fn as_raw_fd(&self) -> RawFd {
[00:57:38] 146 | |         self.as_inner().fd().raw()
[00:57:38] 147 | |     }
[00:57:38] 148 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::AsRawFd` for type `process::ChildStdout`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:161:1
[00:57:38]     |
[00:57:38] 161 | / impl AsRawFd for process::ChildStdout {
[00:57:38] 162 | |     fn as_raw_fd(&self) -> RawFd {
[00:57:38] 163 | |         self.as_inner().fd().raw()
[00:57:38] 164 | |     }
[00:57:38] 165 | | }
[00:57:38]     | |_^ conflicting implementation for `process::ChildStdout`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 151 | / impl AsRawFd for process::ChildStdout {
[00:57:38] 152 | |     fn as_raw_fd(&self) -> RawFd {
[00:57:38] 153 | |         self.as_inner().fd().raw()
[00:57:38] 154 | |     }
[00:57:38] 155 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::AsRawFd` for type `process::ChildStderr`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:168:1
[00:57:38]     |
[00:57:38] 168 | / impl AsRawFd for process::ChildStderr {
[00:57:38] 169 | |     fn as_raw_fd(&self) -> RawFd {
[00:57:38] 170 | |         self.as_inner().fd().raw()
[00:57:38] 171 | |     }
[00:57:38] 172 | | }
[00:57:38]     | |_^ conflicting implementation for `process::ChildStderr`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 158 | / impl AsRawFd for process::ChildStderr {
[00:57:38] 159 | |     fn as_raw_fd(&self) -> RawFd {
[00:57:38] 160 | |         self.as_inner().fd().raw()
[00:57:38] 161 | |     }
[00:57:38] 162 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::FromRawFd` for type `process::Stdio`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:145:1
[00:57:38]     |
[00:57:38] 145 | / impl FromRawFd for process::Stdio {
[00:57:38] 146 | |     unsafe fn from_raw_fd(fd: RawFd) -> process::Stdio {
[00:57:38] 147 | |         let fd = sys::fd::FileDesc::new(fd);
[00:57:38] 148 | |         let io = sys::process::Stdio::Fd(fd);
[00:57:38] 149 | |         process::Stdio::from_inner(io)
[00:57:38] 150 | |     }
[00:57:38] 151 | | }
[00:57:38]     | |_^ conflicting implementation for `process::Stdio`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 135 | / impl FromRawFd for process::Stdio {
[00:57:38] 136 | |     unsafe fn from_raw_fd(fd: RawFd) -> process::Stdio {
[00:57:38] 137 | |         let fd = sys::fd::FileDesc::new(fd);
[00:57:38] 138 | |         let io = sys::process::Stdio::Fd(fd);
[00:57:38] 139 | |         process::Stdio::from_inner(io)
[00:57:38] 140 | |     }
[00:57:38] 141 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::IntoRawFd` for type `process::ChildStdin`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:175:1
[00:57:38]     |
[00:57:38] 175 | / impl IntoRawFd for process::ChildStdin {
[00:57:38] 176 | |     fn into_raw_fd(self) -> RawFd {
[00:57:38] 177 | |         self.into_inner().into_fd().into_raw()
[00:57:38] 178 | |     }
[00:57:38] 179 | | }
[00:57:38]     | |_^ conflicting implementation for `process::ChildStdin`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 165 | / impl IntoRawFd for process::ChildStdin {
[00:57:38] 166 | |     fn into_raw_fd(self) -> RawFd {
[00:57:38] 167 | |         self.into_inner().into_fd().into_raw()
[00:57:38] 168 | |     }
[00:57:38] 169 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::IntoRawFd` for type `process::ChildStdout`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:182:1
[00:57:38]     |
[00:57:38] 182 | / impl IntoRawFd for process::ChildStdout {
[00:57:38] 183 | |     fn into_raw_fd(self) -> RawFd {
[00:57:38] 184 | |         self.into_inner().into_fd().into_raw()
[00:57:38] 185 | |     }
[00:57:38] 186 | | }
[00:57:38]     | |_^ conflicting implementation for `process::ChildStdout`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 172 | / impl IntoRawFd for process::ChildStdout {
[00:57:38] 173 | |     fn into_raw_fd(self) -> RawFd {
[00:57:38] 174 | |         self.into_inner().into_fd().into_raw()
[00:57:38] 175 | |     }
[00:57:38] 176 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error[E0119]: conflicting implementations of trait `sys::unix_ext::io::IntoRawFd` for type `process::ChildStderr`:
[00:57:38]    --> /checkout/src/libstd/sys/unix/ext/process.rs:189:1
[00:57:38]     |
[00:57:38] 189 | / impl IntoRawFd for process::ChildStderr {
[00:57:38] 190 | |     fn into_raw_fd(self) -> RawFd {
[00:57:38] 191 | |         self.into_inner().into_fd().into_raw()
[00:57:38] 192 | |     }
[00:57:38] 193 | | }
[00:57:38]     | |_^ conflicting implementation for `process::ChildStderr`
[00:57:38]     | 
[00:57:38]    ::: /checkout/src/libstd/sys/redox/ext/process.rs
[00:57:38]     |
[00:57:38] 179 | / impl IntoRawFd for process::ChildStderr {
[00:57:38] 180 | |     fn into_raw_fd(self) -> RawFd {
[00:57:38] 181 | |         self.into_inner().into_fd().into_raw()
[00:57:38] 182 | |     }
[00:57:38] 183 | | }
[00:57:38]     | |_- first implementation here
[00:57:38] 
[00:57:38] error: Compilation failed, aborting rustdoc
[00:57:38] 
[00:57:38] [m[m[31m[1merror:[m Could not document `std`.
```

</details>

File: src/libstd/sys/mod.rs [hunk 1]
```diff
@@ -32,49 +32,66 @@
...
+    }
+}
+
+// Import essential modules from both platforms when documenting. These are
+// then later used in the `std::os` module when documenting, for example,
+// Windows when we're compiling for Linux.
+
+#[cfg(dox)]
+cfg_if! {
+    if #[cfg(unix)] {
```
Comment (kennytm) [inline src/libstd/sys/mod.rs]: This should be `#[cfg(any(unix, target_os = "redox"))]` I guess.

**Comment by huangjj27** [comment]
I found that function with return type `String`, `Vec<i32>` or `&'static str` can't be recognized by the front end, the function called return `undefined`. Is it some part of unfinshed work that cause this issue?

**Comment by badboy** [comment]
@huangjj27 That's on purpose and is simply the way Rust works. Strings, Vec, slices, etc. are returned by writing a pointer and additional data to a caller-allocated memory buffer, from which the caller has to read it after calling the function. Wasm has no idea about the layout of Strings. @killercup and I are working on a bit of JS boilerplate to handle those cases easily in [wasm-experiments](https://github.com/killercup/wasm-experiments)

**Comment by kennytm** [comment]
@huangjj27 See also https://stackoverflow.com/questions/47529643/how-to-return-a-string-or-similar-from-rust-in-webassembly

**Pull request merged**